lib/galaxy/webapps/galaxy/services/history_contents.py

```diff
@@ -1133,86 +1172,100 @@ def __get_accessible_collection(self, trans, id: DecodedDatabaseIdField):
             id=id,
         )
 
+    def _create_datasets_from_library_folder(
+        self,
+        trans,
+        history: History,
+        payload: CreateHistoryContentPayloadFromCopy,
+    ):
+        rval = []
+        content = payload.content
+        if content is None:
+            raise exceptions.RequestParameterMissingException("'content' id of lda or hda is missing")
+
+        folder_id = content
+        folder = self.folder_manager.get(trans, folder_id)
+
+        current_user_roles = trans.get_current_user_roles()
+        security_agent: GalaxyRBACAgent = trans.app.security_agent
+
+        def traverse(folder):
+            admin = trans.user_is_admin
+            rval = []
+            for subfolder in folder.active_folders:
+                if not admin:
+                    can_access, folder_ids = security_agent.check_folder_contents(
+                        trans.user, current_user_roles, subfolder
+                    )
+                if (admin or can_access) and not subfolder.deleted:
+                    rval.extend(traverse(subfolder))
+            for ld in folder.datasets:
+                if not admin:
+                    can_access = security_agent.can_access_dataset(
+                        current_user_roles, ld.library_dataset_dataset_association.dataset
+                    )
+                if (admin or can_access) and not ld.deleted:
+                    rval.append(ld)
+            return rval
+
+        for ld in traverse(folder):
+            hda = ld.library_dataset_dataset_association.to_history_dataset_association(history, add_to_history=True)
```

Comment:
You're committing once per added item here, that's what I meant when I said batch import could use some improvements.

Reply:
https://github.com/galaxyproject/galaxy/pull/18970